ledger/complete/mtrie/trieCache.go

```diff
@@ -43,23 +44,21 @@ func (tc *TrieCache) Purge() {
 	tc.lock.Lock()
 	defer tc.lock.Unlock()
 
-	if tc.count == 0 {
+	if tc.tries.Count() == 0 {
 		return
 	}
 
-	toEvict := 0
-	for i := 0; i < tc.capacity; i++ {
-		toEvict = (tc.tail + i) % tc.capacity
-		if tc.onTreeEvicted != nil {
-			if tc.tries[toEvict] != nil {
-				tc.onTreeEvicted(tc.tries[toEvict])
-			}
-		}
-		tc.tries[toEvict] = nil
+	toEvict := tc.tries.Tries()
+
+	// update the state first
+	capacity := tc.tries.Capacity()
+	tc.tries = common.NewTrieQueue(capacity, tc.onTreeEvicted)
+	tc.lookup = make(map[ledger.RootHash]*trie.MTrie, capacity)
+
+	// evicting all nodes after state is reset
+	for _, trie := range toEvict {
+		tc.onTreeEvicted(trie)
```

Comment:
I don't like this onTrieEviceted function:
Is is called before the lock is released, if someone is trying to report some metrics by calling `TrieCache.Count()`, it will introduce a dead lock by mistake.

A better pattern would be just return the evicted items to the caller _after_ releasing the lock, that's the moment actually the item is actually evicted.

Same idea for the `Push` method, it could also return the evicted item in the return value after the lock is released. It avoid chance of introducing dead lock if misused.

Is any code using this callback? I didn't find any